File: Java/src/gov/nasa/larcfm/ACCoRD/CD3DTable.java
/*
 * Copyright (c) 2014-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.ACCoRD;

import java.util.HashMap;
import gov.nasa.larcfm.Util.ParameterData;
import gov.nasa.larcfm.Util.Units;
import gov.nasa.larcfm.Util.f;

public class CD3DTable implements ParameterTable {
	double D;
	double H;

	private HashMap<String,String> units_;

	public CD3DTable() {
		D = Units.from("nmi",5);
		H = Units.from("ft",1000);
		units_ = new HashMap<String,String>();
		units_.put("D","nmi");
		units_.put("H","ft");
	}

	public CD3DTable(double d, double h) {
		D = Math.abs(d);
		H = Math.abs(h);
		units_ = new HashMap<String,String>();
		units_.put("D","m");
		units_.put("H","m");
	}

	public CD3DTable(double d, String dunit, double h, String hunit) {
		D = Units.from(dunit, Math.abs(d));
		H = Units.from(hunit, Math.abs(h));
		units_ = new HashMap<String,String>();
		units_.put("D",dunit);
		units_.put("H",hunit);
	}

	/**
	 * Copy constructor -- returns a fresh copy.
	 */
	public CD3DTable(CD3DTable tab) {
		units_ = new HashMap<String,String>();
		copyValues(tab);
	}

	public CD3DTable copy() {
		return new CD3DTable(this);
	}

	public void copyValues(CD3DTable tab) {
		D = tab.D;
		H = tab.H;
		units_.clear();
		units_.putAll(tab.units_);
	}

	public double getHorizontalSeparation() {
		return D;
	}

	public void setHorizontalSeparation(double d) {
		D = Math.abs(d);
	}

	public double getVerticalSeparation() {
		return H;
	}

	public void setVerticalSeparation(double h) {
		H = Math.abs(h);
	}

	public double getHorizontalSeparation(String u) {
		return Units.to(u, D);
	}

	public void setHorizontalSeparation(double d, String u) {
		setHorizontalSeparation(Units.from(u,Math.abs(d)));
		units_.put("D",u);
	}

	public double getVerticalSeparation(String u) {
		return Units.to(u, H);
	}

	public void setVerticalSeparation(double h, String u) {
		setVerticalSeparation(Units.from(u,Math.abs(h)));
		units_.put("H",u);
	}

	public ParameterData getParameters() {
		ParameterData p = new ParameterData();
		updateParameterData(p);
		return p;
	}

	public void updateParameterData(ParameterData p) {
	  p.setInternal("D",D,getUnits("D"));
	  p.setInternal("H",H,getUnits("H"));
	}

	public void setParameters(ParameterData p) {
		if (p.contains("D")) {
			setHorizontalSeparation(p.getValue("D"));
			units_.put("D",p.getUnit("D"));
		}
		if (p.contains("H")) {
			setVerticalSeparation(p.getValue("H"));
			units_.put("H",p.getUnit("H"));
		}
	}

	public String getUnits(String key) {
		String u = units_.get(key);
		if (u == null) {
			return "unspecified";
		}
		return u;
	}

	public String toString() {
		return "D = "+Units.str(getUnits("D"),D)+", H = "+Units.str(getUnits("H"),H);
	}

	public String toPVS(int prec) {
		return "(# D:= "+f.FmPrecision(D,prec)+", H:= "+f.FmPrecision(H,prec)+" #)";
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		long temp;
		temp = Double.doubleToLongBits(D);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		temp = Double.doubleToLongBits(H);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		CD3DTable other = (CD3DTable) obj;
		if (Double.doubleToLongBits(D) != Double.doubleToLongBits(other.D))
			return false;
		if (Double.doubleToLongBits(H) != Double.doubleToLongBits(other.H))
			return false;
		return true;
	}

}

File: C++/src/ACCoRD/CD3DTable.cpp
/*
 * Copyright (c) 2014-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
#include "Units.h"
#include "CD3DTable.h"
#include "ParameterData.h"
#include "format.h"

namespace larcfm {

CD3DTable::CD3DTable() {
  D = Units::from("nmi", 5.0);
  H = Units::from("ft", 1000.0);
  units_["D"] = "nmi";
  units_["H"] = "ft";
}

CD3DTable::CD3DTable(double d, double h) {
  D = std::abs(d);
  H = std::abs(h);
  units_["D"] = "m";
  units_["H"] = "m";
}

CD3DTable::CD3DTable(double d, const std::string& dunit, double h, const std::string& hunit) {
  D = Units::from(dunit, std::abs(d));
  H = Units::from(hunit, std::abs(h));
  units_["D"] = dunit;
  units_["H"] = hunit;
}

/**
 * Copy constructor -- returns a fresh copy.
 */
CD3DTable::CD3DTable(const CD3DTable& tab) {
  copyValues(tab);
}

CD3DTable CD3DTable::copy() const {
  CD3DTable tab;
  tab.D = D;
  tab.H = H;
  tab.units_ = units_;
  return tab;
}

void CD3DTable::copyValues(const CD3DTable& tab) {
  D = tab.D;
  H = tab.H;
  units_ = tab.units_;
}

double CD3DTable::getHorizontalSeparation() const {
  return D;
}

void CD3DTable::setHorizontalSeparation(double d) {
  D = std::abs(d);
}

double CD3DTable::getVerticalSeparation() const {
  return H;
}

void CD3DTable::setVerticalSeparation(double h) {
  H = std::abs(h);
}

double CD3DTable::getHorizontalSeparation(const std::string& u) const {
  return Units::to(u, D);
}

void CD3DTable::setHorizontalSeparation(double d, const std::string& u) {
  setHorizontalSeparation(Units::from(u,d));
  units_["D"] = u;
}

double CD3DTable::getVerticalSeparation(const std::string& u) const {
  return Units::to(u, H);
}

void CD3DTable::setVerticalSeparation(double h, const std::string& u) {
  setVerticalSeparation(Units::from(u,h));
  units_["H"] = u;
}

ParameterData CD3DTable::getParameters() const {
  ParameterData p;
  updateParameterData(p);
  return p;
}

void CD3DTable::updateParameterData(ParameterData& p) const {
  p.setInternal("D",D,getUnits("D"));
  p.setInternal("H",H,getUnits("H"));
}

void CD3DTable::setParameters(const ParameterData& p) {
  if (p.contains("D")) {
    setHorizontalSeparation(p.getValue("D"));
    units_["D"] = p.getUnit("D");
  }
  if (p.contains("H")) {
    setVerticalSeparation(p.getValue("H"));
    units_["H"] = p.getUnit("H");
  }
}

std::string CD3DTable::getUnits(const std::string& key) const {
  std::map<std::string,std::string>::const_iterator got = units_.find(key);
  if (got == units_.end()) {
    return "unspecified";
  }
  return got->second;
}

bool CD3DTable::equals(const CD3DTable& t2) const {
  return D == t2.D && H == t2.H;
}

std::string CD3DTable::toString() const {
  return "D = "+Units::str(getUnits("D"),D)+", H = "+Units::str(getUnits("H"),H);
}

std::string CD3DTable::toPVS(int prec) const {
  return "(# D:= "+FmPrecision(D,prec)+", H:= "+FmPrecision(H,prec)+" #)";
}

}
